WordPress/gutenberg
# Fix multiple tooltips showing on NavigableToolbars

The 'core/block-editor/focus-toolbar' shortcut (alt + F10) selects the active navigable toolbar's first tabbable item, but when there are multiple navigable toolbars, it places focus on all of the active navigable toolbar components. This results in an odd state of two tooltips being open. This commit scaffolds the ability to disable the keyboard shortcut from applying.



<img width="1009" alt="Editor screen with Toggle block inserter and Paragraph tooltips open and focus on both buttons" src="https://user-images.githubusercontent.com/967608/230458867-3e3d2041-e971-4fea-9e31-41e1e61f65f1.png">

TODO: Figure out how to identify if multiple `NavigableToolbars` are present, or at least tell the document toolbar not to show if the block toolbar is present.

## What?
Only apply the shortcut to one `NavigableToolbar` component.

## Why?
Each `NavigableToolbar` gets an `alt + F10` (or `fn + option + F10` on mac keyboards) shortcut added to it. When there are multiple `NavigableToolbar`s components on the page, they all try to focus the first tabbable element within it, displaying multiple popovers at once.

## How?
Add `useKeyboardShortcut` prop to `NavigableToolbar` and `useToolbarFocus` components so the document `HeaderToolbar` can disable itself from using the shortcut if the block toolbar is also showing. 

If there are no selected blocks OR we're not in `edit` mode, then we'll allow the keyboard shortcut for the header toolbar to be active. There may be other conditions I'm missing to identify if a focusable block toolbar is showing.

The shortcut says `Navigate to the nearest toolbar` which means to me, "Move to the toolbar closest to my current focus in the DOM," but the implementation here is, "Move to the most specific toolbar." As in, if there is a block toolbar open, move to it. If there is only one toolbar open, move to it.

## Testing Instructions
These are all documented in the e2e tests. To see them all running, add:

```
test.use( {
	launchOptions: {
		slowMo: 500,
	},
} );
``` 

to the top of `test/e2e/specs/editor/various/shortcut-focus-toolbar.spec.js` (like, line 5) and then run these tests with `npm run test:e2e:playwright -- shortcut-focus-toolbar
 --headed`



### Testing Instructions for Keyboard

#### Block toolbar open
- Move focus into a block to display the block toolbar. 
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be applied to the block toolbar

#### No selected blocks
- Hide the block toolbar by moving focus to the editor page title (or anywhere else)
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be in the document toolbar 

#### In Select Mode
- In the top document toolbar, change to Select mode
- Place focus on a block
- Press `alt + F10` (or `fn + option + F10` on mac keyboards)
- Focus should be in the document toolbar 

Basically, try to break it using all of the different view modes (Top Toolbar on/off, Spotlight mode, etc) and with an empty default blocks selected, a block selected, multiple blocks selected, etc. Anytime a block toolbar is showing, it should receive focus with the `alt + F10`  shortcut, and anytime it's _not_ showing, focus should be in the top toolbar.

## Screenshots or screencast 
**Broken state:**
https://user-images.githubusercontent.com/967608/230490886-5849ba0e-1910-47b0-b522-750b2f99e70c.mov
